rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,191 @@
+the metric to determine how to process the tags.
+
+
+## Doc-level Proposal
+
+The Remap language can work with the metadata around a metric.
+
+The following functions are provided:
+
+### `get_name() -> string`
+Returns the name of the metric.
+
+### `set_name(string)`
+Allows you to change the name of the metric.
+
+### `get_namespace() -> string`
+Returns the namepace of the metric.
+
+### `set_namespace(string)`
+Allows you to change the namespace of the metric.
+
+### `get_timestamp() -> timestamp`
+Returns the timestamp of the metric.
+
+### `set_timestamp(timestamp)`
+Allows you to change the timestamp of the metric.
+
+### `get_kind() -> string`
+Returns the kind of the metric, will either be "Incremental" or "Absolute".
+
+### `set_kind(string)`
+Allows you to set the kind of the metric. Any string other than "Incremental" or
+"Absolute" will result in an error.
+
+### `get_tag(string) -> string`
+Returns the value of the given tag.
+
+### `add_tag(string, string)`
+Allows you to set the value of a given tag.
+
+### `rename_tag(string, string)`
+Rename the tag with the given key to the new key.
+
+### `remove_tag(string)`
+Removes the tag with the given key from the metric.
+
+Note that accessing path values (identifiers starting with a `.`) are not 
+available for metric events.
+
+Variables (identifiers starting with a `$`) are available.
+
+## Rationale
+
+Enabling the Remap language to work with metric events will allow users to
+simplify their configurations if they already use the `*_tag` transforms
+provided. It will also allow them remove the use of the `Lua` transform if they
+have been falling back to that for anything but the most complex manipulation.
+
+  
+## Prior Art
+
+
+## Drawbacks
+
+In order to maintain these functions there will be a slight maintenance burden.
+
+By exposing the metric data to Remap, it may place some restrictions should we 
+want to change the internal model in the future. The more we decide to expose to
+the language, the more it may restrict us going forwards if we need to ensure
+backwards compatibility for users configurations.
+
+## Alternatives
+
+### Use paths
+
+We could allow the metric data to be manipulated through paths in a similar way
+to how the language works with log events.
+
+The following fields would be available:
+
+- `.name` - The name of the metric.
+- `.namespace` - The namespace for the metric.
+- `.timestamp` - Timestamp of the metric.
+- `.kind` - The kind of the metric - *Incremental* or *Absolute*
+- `.tags` - A map containing the tags set for the metric.
+ 
+However, there are strict limitations to how this data can be represented. 
+Setting `name` or `namespace` to anything other than a string should be an 
+error. `tags` has to be a map of string keys to string values. No nesting should
+be allowed. The inflexibility around this data means that the attempt to make
+it "feel" the same as log events would not be effective.
```

Comment:
I feel like this is the biggest discussion point of this RFC.

Given the current proposal, metrics feel too much like second-class citizens, or a tacked-on capability of the Remap language.

When someone new is introduced to Remap, it is immediately clear that the language is designed around log data, whereas metrics "works, but isn't nearly as seamless".

I would actually _prefer_ if we used paths to manipulate metrics, as opposed to the proposed functions.

My proposal (hand-wavy on implementation details) would be as follows:

* We extend Remap to allow us to define which paths can be modified/queried, and what values can be assigned to them.
* For `log` events, we set this restriction to "any path, any value"
* For `metrics` events, we set the list of paths you mentioned above, and define the values that you are allowed to assign to these paths.
* This allows us to provide boot-time errors for invalid scripts that try to manipulate metrics (e.g. `.tags = "foo"` fails, but `.tags.my_tag = "foo"` is allowed, etc)

Deleting a tag can be done using the existing `del` function, and other functions will be available as well (when it makes sense).

Doing this, we don't need any of the proposed functions, but still keep tight control on what can be manipulated in the metrics events.

Reply:
In all honesty, I went back and forth between paths and functions more times than I could count! I settled on functions in the end because there are so many invalid things that the user can do with paths, and there is no precise way of specifying what is and isn't invalid beyond the documentation and friendly error messages. With functions, it is instantly apparent what can and can't be done.

Metrics are, by design, a more rigid data structure than logging information. The user is going to expect to have less flexibility in manipulating them than with logs.

They still have variables, which they are completely free to use however they need.  I think by suggesting they can use paths, but then heavily restricting their use will make metrics feel more like a second class citizen than if we just reserved paths purely for logging.

Reply:
> there is no precise way of specifying what is and isn't invalid beyond the documentation and friendly error messages

I would disagree with this statement, as I've demonstrated in my (admeditly limited in scope) proposal.

> Metrics are, by design, a more rigid data structure than logging information. The user is going to expect to have less flexibility in manipulating them than with logs.

I concur with this, which is why I believe guiding a user through both documentation and boot-time errors will be of help here.

>  I think by suggesting they can use paths, but then heavily restricting their use will make metrics feel more like a second class citizen than if we just reserved paths purely for logging.

At first, the paths are restricted, in that they only allow manipulating a fixed set of metadata fields (the ones you listed), but once you open up the `.metrics` path, it becomes easier to add new metrics or manipulate existing metrics through paths as well.

---

I get where you are coming from, and I'm also not 100% sold on the path-based solution, but I'm even less sold on the proposed set of functions.

I suspect having a couple of real-world use-cases (ranging from simple to complex), and writing those up using functions and paths would help us better understand the implications of both solutions in different scenarios.

Reply:
There's no technical reason why we couldn't implement both.

I agree, some real world use cases would be helpful.

Reply:
> There's no technical reason why we couldn't implement both.

You mean for testing purposes or as a final solution?

I hope the RFC and this discussion allows us to come to a final solution, but either way, I don't think we should have both implementations as a final solution. I'm not one to rigorously promote "one true way" for anything, but I do believe we should commit to one way to work with events, not multiple, unless there are clear upsides to doing so (which I don't believe there are right now).

Reply:
I do mean as a final solution. I'm not necessarily saying it's a good idea! But it is definitely possible and shouldn't be discounted straight away.

Reply:
> I do mean as a final solution. I'm not necessarily saying it's a good idea! But it is definitely possible and shouldn't be discounted straight away.

Got it 👍

I personally _would_ discount such a solution because of the added learning curve it adds that can be avoided by resolving this discussion towards the one solution we like most.

But you are correct that there is no technical reason we can't support both, and maybe others feel differently.